File: src/test/java/no/nav/pensjondokdist/saf/SafClientTest.java
package no.nav.pensjondokdist.saf;

import static org.assertj.core.api.Assertions.assertThat;
import static org.springframework.test.web.client.ExpectedCount.manyTimes;
import static org.springframework.test.web.client.match.MockRestRequestMatchers.requestTo;
import static org.springframework.test.web.client.response.MockRestResponseCreators.withBadRequest;
import static org.springframework.test.web.client.response.MockRestResponseCreators.withSuccess;

import java.io.IOException;

import com.fasterxml.jackson.core.type.TypeReference;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.web.client.MockRestServiceServer;
import org.springframework.web.client.RestTemplate;

import no.nav.pensjondokdist.PensjonDokdistException;
import no.nav.pensjondokdist.oidc.OidcProvider;
import no.nav.pensjondokdist.saf.graphql.GraphQlError;
import no.nav.pensjondokdist.saf.graphql.GraphQlResponse;
import no.nav.pensjondokdist.util.JsonUtil;

@SpringBootTest
@RunWith(MockitoJUnitRunner.class)
public class SafClientTest {
    private static final String URL = "http://SafUrl.com";
    private SafClient client;
    @Mock
    private OidcProvider provider;
    private RestTemplate restTemplate = new RestTemplate();

    @Before
    public void init() throws IOException {
        client = new SafClient(URL, restTemplate, provider);
   }

    @Test
    public void testGetWithSuccessResponse() {
        MockRestServiceServer server = MockRestServiceServer.bindTo(restTemplate).build();

        server.expect(manyTimes(), requestTo(URL))
                .andRespond(withSuccess(getJsonString(), MediaType.APPLICATION_JSON));
        client.fetchJournal("12345");
        server.verify();
    }

    @Test(expected = PensjonDokdistException.class)
    public void testGetWithNullBodyShouldThrowException() {
        MockRestServiceServer server = MockRestServiceServer.bindTo(restTemplate).build();
        server.expect(manyTimes(), requestTo(URL))
                .andRespond(withSuccess("", MediaType.APPLICATION_JSON));
        client.fetchJournal("12345");
        server.verify();
    }

    @Test(expected = PensjonDokdistException.class)
    public void testGetWithBadRequestResponseThrowException() {
        MockRestServiceServer server = MockRestServiceServer.bindTo(restTemplate).build();
        server.expect(manyTimes(), requestTo(URL))
                .andRespond(withBadRequest());
        client.fetchJournal("12345");
        server.verify();
    }

    @Test
    public void testErrorResponse() {
        GraphQlResponse graphQlResponse = JsonUtil.toObjectFromJsonString(getErrorString(),  new TypeReference<GraphQlResponse>(){});
        assertThat(graphQlResponse).isNotNull();
        assertThat(graphQlResponse.getErrors()).hasSize(1);
        GraphQlError error = graphQlResponse.getErrors().get(0);
        assertThat(error).isNotNull();
        assertThat(error.getMessage()).isNotNull();
        assertThat(error.getExceptionType()).isEqualTo("safexception");
        assertThat(error.getException()).isEqualTo(NullPointerException.class.getSimpleName());
    }

    private String getErrorString() {
        return "{\n"
                + "  \"errors\": [\n"
                + "    {\n"
                + "      \"message\": \"Feilet ved henting av data (/journalpost) : null\",\n"
                + "      \"locations\": [\n"
                + "        {\n"
                + "          \"line\": 3,\n"
                + "          \"column\": 3\n"
                + "        }\n"
                + "      ],\n"
                + "      \"path\": [\n"
                + "        \"journalpost\"\n"
                + "      ],\n"
                + "      \"exceptionType\": \"safexception\",\n"
                + "      \"exception\": \"NullPointerException\"\n"
                + "    }\n"
                + "  ],\n"
                + "  \"data\": {\n"
                + "    \"journalpost\": null\n"
                + "  }\n"
                + "}";
    }
    private String getJsonString() {
        return "{\"data\":{\"journalpost\":{\"journalpostId\":\"453501014\",\"journalforendeEnhet\":\"4803\"}}}";
    }
}

File: src/test/kotlin/no/nav/pensjon/dokdist/saf/SafClientTest.kt
package no.nav.pensjon.dokdist.saf

import com.fasterxml.jackson.databind.ObjectMapper
import no.nav.pensjon.dokdist.auth.OnBehalfOfTokenResponse
import no.nav.pensjon.dokdist.graphql.GraphQLResponse
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.*
import org.springframework.boot.test.web.client.MockServerRestTemplateCustomizer
import org.springframework.boot.web.client.RestTemplateBuilder
import org.springframework.http.*
import org.springframework.test.web.client.match.MockRestRequestMatchers.*
import org.springframework.test.web.client.response.MockRestResponseCreators.*
import org.springframework.web.server.ResponseStatusException

private const val endpoint = "http://saf.local"

class SafClientTest {
    private val objectMapper = ObjectMapper()

    private val token = OnBehalfOfTokenResponse("Bearer", "their-scope", "1000", "the-token")
    private val journalpost = Journalpost("12345", "en-kul-enhet", listOf(Journalpost.Dokument("kult-brev")))
    private val responseJson = objectMapper.writeValueAsString(GraphQLResponse(mapOf("journalpost" to journalpost), null))
    private val graphqlUrl = "$endpoint/graphql"

    private val mockServerCustomizer: MockServerRestTemplateCustomizer = MockServerRestTemplateCustomizer()
    private val saf = SafClient(endpoint, token, RestTemplateBuilder(mockServerCustomizer))
    private val mockServer = mockServerCustomizer.server

    @Test
    fun `can fetch journalpost`() {
        mockServer.expect(requestTo(graphqlUrl))
            .andExpect(method(HttpMethod.POST))
            .andExpect(jsonPath(".variables.journalpostId").value(journalpost.journalpostId))
            .andExpect(jsonPath(".query").exists())
            .andRespond(withSuccess(responseJson, MediaType.APPLICATION_JSON))

        saf.fetchJournal(journalpost.journalpostId)
        mockServer.verify()
    }

    @Test
    fun `missing journalpost with notfound status returns null`() {
        mockServer.expect(requestTo(graphqlUrl))
            .andRespond(withStatus(HttpStatus.NOT_FOUND))

        assertThat(saf.fetchJournal(journalpost.journalpostId)).isNull()
        mockServer.verify()
    }

    @Test
    fun `missing journalpost with ok status returns null`() {
        val graphQLResponse = GraphQLResponse(mapOf("journalpost" to null), listOf(GraphQLResponse.Error("Fant ikke journalpost", null, null, null, null)))
        mockServer.expect(requestTo(graphqlUrl))
            .andRespond(withSuccess(objectMapper.writeValueAsString(graphQLResponse), MediaType.APPLICATION_JSON))

        assertThat(saf.fetchJournal(journalpost.journalpostId)).isNull()
        mockServer.verify()
    }

    @Test
    fun `empty response fails`() {
        mockServer.expect(requestTo(graphqlUrl))
            .andRespond(withSuccess())

        assertThrows<SafException> { saf.fetchJournal(journalpost.journalpostId) }
        mockServer.verify()
    }

    @Test
    fun `graphql error fails`() {
        val response = GraphQLResponse(null, listOf(GraphQLResponse.Error("Some syntax error or something", null, null, null, null)))
        mockServer.expect(requestTo(graphqlUrl))
            .andRespond(withSuccess(objectMapper.writeValueAsString(response), MediaType.APPLICATION_JSON))

        assertThrows<SafException> { saf.fetchJournal(journalpost.journalpostId) }
        mockServer.verify()
    }

    @Test
    fun `forbidden is rethrown as responseStatusException`() {
        mockServer.expect(requestTo(graphqlUrl))
            .andRespond(withStatus(HttpStatus.FORBIDDEN))

        val exception = assertThrows<ResponseStatusException> { saf.fetchJournal(journalpost.journalpostId) }
        assertThat(exception).matches { it.statusCode == HttpStatus.FORBIDDEN }
        mockServer.verify()
    }

    @Test
    fun `bad request fails`() {
        mockServer.expect(requestTo(graphqlUrl))
            .andRespond(withBadRequest())

        assertThrows<SafException> { saf.fetchJournal(journalpost.journalpostId) }
        mockServer.verify()
    }

    @Test
    fun `server error fails`() {
        mockServer.expect(requestTo(graphqlUrl))
            .andRespond(withServerError())

        assertThrows<SafException> { saf.fetchJournal(journalpost.journalpostId) }
        mockServer.verify()
    }

}
